### ISSUE TITLE
Implement `WinRateCallback` to guide model evaluation during training

### ISSUE THREAD

**Issue by lewtun** [huggingface/trl#1397]
Evaluating chat / instruct models is _hard_ and relying on proxy metrics like loss / reward / accuracy doesn't typically correlate well with downstream benchmarks like AlpacaEval / IFEval etc.

To help guide training, it would be good to have a `WinRateCallback` that can be passed to the various trainers and does the following:

1. On a set of user-specified prompts, computes completions from the reference model or initial model before any optimisation steps. These completions could also be provided in advance with the prompt dataset.
2. During each evaluation step, the callback generates completions from the active model.
3. A judge LLM or reward model ranks the completion pairs to compute the win rate of the active model vs the reference model.

We could also log the win rates in WandB together with the completion lengths (which are correlated with judge LLM scores) and visualise them like the following examples from [this](https://arxiv.org/abs/2402.04833) and [this](https://arxiv.org/abs/2402.04792):

![Screenshot 2024-03-05 at 13 14 07](https://github.com/huggingface/trl/assets/26859204/3237134e-f425-4089-a5dd-eda26a668ba5)

![Screenshot 2024-03-05 at 13 15 25](https://github.com/huggingface/trl/assets/26859204/48b2e560-926d-4fe7-830c-0b5be5b15a13)

Concretely, I think we want something like the following:

```python
class WinRateCallback(TrainerCallback):
    def __init__(self, prompt_dataset: Dataset, prompt_column: str, generation_config: GenerationConfig, judge_model):
        self.prompts = prompt_dataset[prompt_column]
        self.generation_config = generation_config
        self.completions = []

    def on_evaluate(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
        model = kwargs["model"]
        tokenizer = kwargs["tokenizer"]
        distributed_state = PartialState()
        model.to(distributed_state.device)
        model.eval()  # Set model to evaluation mode
        completions_per_process = []
        # Generalise this to do batched inference
        with distributed_state.split_between_processes(self.prompts, apply_padding=True) as prompts:
            for message in tqdm(prompts, desc="Generating completions"):
                tokenized_message = tokenizer.apply_chat_template(
                    [message], tokenize=True, add_generation_prompt=True, return_tensors="pt"
                ).to(model.device)
                output = model.generate(tokenized_message, self.generation_config)
                # Slice out prompt tokens and decode
                generated_text = tokenizer.decode(output[0][tokenized_message.shape[-1] :], skip_special_tokens=True)
                completions_per_process.append(generated_text)

        # Gather completions across all processes
        completions_gather = [None for _ in range(dist.get_world_size())]
        dist.all_gather_object(completions_gather, completions_per_process)
        # Flatten data
        completions = [msg for sublist in completions_gather for msg in sublist]
        # Drop duplicates produced by padding
        self.completions = completions[: len(self.prompts)]

       # Rank and compute win rate etc etc
      scores = judge_model(prompts, completions, reference_completions)
```

**Comment by lewtun** [comment]
GitHub bot begone!

**Comment by AmanSinghal927** [comment]
Thanks for sharing this!

**Comment by ruggsea** [comment]
I wholeheartedly support this, I was shocked to see how much trl focuses on simplifying finetuning but at the same time not enabling more experimentation with evals (not yet at least). I can think of various different modes of this callback that could be implemented:
- the one @lewtun proposes that pairwise compares completions with reference ones
- one in which the judge-llm rates on a certain scale the completion quality (based on some parameters that could be passed as arguments to be injected in the evaluation prompt) (for reference: [this preprint](https://arxiv.org/pdf/2305.13711) or this [HF cookbook](https://www.databricks.com/blog/LLM-auto-eval-best-practices-RAG)) 
- one akin to the first one, but the reference completions are taken from generations of the previous checkpoints of the model: this way, the judge-llm could indicate whether the model is still improving or not
